File: LeetCode/Medium/207. Course Schedule/Course Schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int N = numCourses;
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < N; i++) graph.add(new ArrayList<>());
        int[] inDegree = new int[N + 1];

        for (int[] parts : prerequisites) {
            int v = parts[0];
            int u = parts[1];
            
            graph.get(u).add(v);
            inDegree[v]++;
        }

        Queue<Integer> q = new LinkedList<>();
        int count = 0;

        for (int i = 0; i < N; i++) {
            if (inDegree[i] == 0) q.offer(i);
        } 

        while (!q.isEmpty()) {
            int u = q.poll();
            count++;

            for (int v : graph.get(u)) {
                inDegree[v]--;
                if (inDegree[v] == 0) q.offer(v);
            }
        }

        return count == N;
    }
}

File: LeetCode/Medium/207. Course Schedule/Course Schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        N = numCourses
        graph = [[] for _ in range(N)]
        in_degree = [0] * N

        for v, u in prerequisites:
            graph[u].append(v)
            in_degree[v] += 1

        q = deque([])
        count = 0

        for i in range(N):
            if in_degree[i] == 0:
                q.append(i)

        while q:
            u = q.popleft()
            count += 1

            for v in graph[u]:
                in_degree[v] -= 1
                if in_degree[v] == 0:
                    q.append(v)
        
        return count == N